pkg/v2/tkr/webhook/cluster/tkr-resolver/cluster/cluster.go

```diff
@@ -161,33 +165,28 @@ func getAnnotation(annotations map[string]string, name string) *string {
 	return &value
 }
 
-// constructOSImageQuery determines if resolution of TKR/OSImage is needed, and creates the *OSImageQuery that
+// filterOSImageQuery determines if resolution of TKR/OSImage is needed. It returns the passed *OSImageQuery that
 // instructs the resolver to perform the resolution. If not, nil is returned, indicating that no resolution is necessary
 // for this particular part of the cluster topology (either controlPlane or a machineDeployment).
-func (cw *Webhook) constructOSImageQuery(v string, tkrSelector, osImageSelector labels.Selector, labelSet labels.Set) *data.OSImageQuery {
-	if tkrName, ok := labelSet[runv1.LabelTKR]; ok {
-		if tkr := cw.TKRResolver.Get(tkrName, &runv1.TanzuKubernetesRelease{}).(*runv1.TanzuKubernetesRelease); tkr != nil {
-			if osImageName, ok := labelSet[runv1.LabelOSImage]; ok {
+func (cw *Webhook) filterOSImageQuery(tkr *runv1.TanzuKubernetesRelease, tkrData TKRData, osImageQuery *data.OSImageQuery) *data.OSImageQuery {
+	if tkr != nil && tkrData != nil {
+		if tkrDataValue := tkrData[tkr.Spec.Kubernetes.Version]; tkrDataValue != nil && tkrDataValue.Labels[runv1.LabelTKR] == tkr.Name {
+			if osImageName, ok := tkrDataValue.Labels[runv1.LabelOSImage]; ok {
 				if osImage := cw.TKRResolver.Get(osImageName, &runv1.OSImage{}).(*runv1.OSImage); osImage != nil {
 					// Found TKR and OSImage. Now, see if they match the provided version and selectors.
-					req, _ := labels.NewRequirement(version.Label(v), selection.Exists, nil)
-					tkrSelectorWithVersion := tkrSelector.Add(*req)
-					osImageSelectorWithVersion := osImageSelector.Add(*req)
-					if tkrSelectorWithVersion.Matches(labels.Set(tkr.Labels)) && osImageSelectorWithVersion.Matches(labels.Set(osImage.Labels)) {
+					if version.Prefixes(tkr.Spec.Version).Has(osImageQuery.K8sVersionPrefix) &&
+						osImageQuery.TKRSelector.Matches(labels.Set(tkr.Labels)) &&
+						osImageQuery.OSImageSelector.Matches(labels.Set(osImage.Labels)) {
 						return nil // indicating we don't need to resolve: already have matching TKR and OSImage
 					}
 				}
 			}
 		}
 	}
-	return &data.OSImageQuery{
-		K8sVersionPrefix: v,
-		TKRSelector:      tkrSelector,
-		OSImageSelector:  osImageSelector,
-	}
+	return osImageQuery
 }
 
-func (cw *Webhook) constructMDQueries(cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass, tkrSelector labels.Selector) ([]*data.OSImageQuery, error) {
+func (cw *Webhook) constructMDQueries(tkr *runv1.TanzuKubernetesRelease, cluster *clusterv1.Cluster, clusterClass *clusterv1.ClusterClass, tkrSelector labels.Selector) ([]*data.OSImageQuery, error) {
```

Comment:
nit: Should this be `constructMDOSImageQueries`?

Reply:
Makes sense.